### PR TITLE
Merge major networking components into dcos-net service

### PR DESCRIPTION
**Pull request by urbanserj**
## High Level Description

DCOS network consist of erlang applications such as Spartan (DNS dual-dispatcher), Minuteman (Layer 4 load balancer), Navstar (Network Util) and Networking-API (Network Metrics). These applications run independently but have a dependency on Lashup (distributed key-value store). They use Lashup as a distributed key-value store and for failure detection. In theory, each application can run their own Lashup instance however, this is undesirable as separate Lashup instances would consume a lot of system resource. An alternate approach could be to run these application together in a single erlang VM so that they all could benefit from single lashup instance.

## Related Issues

  - [DCOS_OSS-1241](https://jira.dcos.io/browse/DCOS_OSS-1241) Merge Spartan and Navstar into dcos-net

## Checklist for all PR's

  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [x] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [x] Test Results: [link to CI job test results for component]
  - [x] Code Coverage (if available): [link to code coverage report]

https://github.com/mesosphere/dcos-net
https://github.com/mesosphere/dcos-net/pull/1
https://github.com/mesosphere/dcos-enterprise/pull/1113

___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**

## Instructions and Review Process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using github's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

### CODE REVIEW

File: docs/admin-ports.md [hunk 1]
```diff
@@ -6,25 +6,24 @@ The following is a list of ports used by internal DC/OS services, and their corr
...
  - 61053: dcos-mesos-dns
  - 61420: dcos-epmd
- - 61421: dcos-minuteman
- - 62053: dcos-spartan
- - 62080: dcos-navstar
- - 62501: dcos-spartan
- - 62502: dcos-navstar
- - 62503: dcos-minuteman
+ - 62053: dcos-net
+ - 62080: dcos-net
```
Comment (drewkerrigan) [inline docs/admin-ports.md]: Maybe clarify each of the dcos-net ports - like `- 62053 dcos-net (dns)` or something similar?

**Comment by urbanserj** [comment]
@mesosphere-mergebot label [Ready For Review]

**Comment by jgehrcke** [comment]
This looks quite good already! Let's have more reviews, please.

I refrain from pushing this forward myself today because it's such a big change. Creating master train 224 right now w/o this.

File: packages/adminrouter/extra/src/docs/api/nginx.master.yaml [hunk 2]
```diff
@@ -234,11 +234,11 @@ routes:
       file: conf/lib/metadata.lua
     path: /metadata
   /navstar/lashup/key:
-    group: Navstar
+    group: DC/OS Net
     matcher: path
     proxy:
       path: 'http://$backend/lashup/key'
-      backend: navstar
+      backend: dcos_net
```
Comment (asridharan) [inline packages/adminrouter/extra/src/docs/api/nginx.master.yaml]: Every where else we are using `dcos-net`, can we s/dcos_net/dcos-net?

File: packages/adminrouter/extra/src/includes/http/master.conf [hunk 3]
```diff
@@ -44,7 +44,7 @@ upstream exhibitor {
     server 127.0.0.1:8181;
 }
 
-# Name: Navstar
-upstream navstar {
+# Name: DC/OS Net
+upstream dcos_net {
```
Comment (asridharan) [inline packages/adminrouter/extra/src/includes/http/master.conf]: ditto s/dcos_net/dcos-net

File: packages/adminrouter/extra/src/includes/server/master.conf [hunk 4]
```diff
@@ -212,14 +212,14 @@ location ~ ^/(slave|agent)/(?<agentid>[0-9a-zA-Z-]+)(?<url>.+)$ {
...
-# Group: Navstar
+# Group: DC/OS Net
 location /navstar/lashup/key {
     access_by_lua_block {
         auth.access_lashupkey_endpoint();
         util.clear_dcos_cookies();
     }
     include includes/proxy-headers.conf;
-    proxy_pass http://navstar/lashup/key;
+    proxy_pass http://dcos_net/lashup/key;
```
Comment (asridharan) [inline packages/adminrouter/extra/src/includes/server/master.conf]: ditto s/docs_net/dcos-net

File: packages/dcos-net/extra/dcos-net-setup.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#!/opt/mesosphere/bin/python
+
+import os
+import subprocess
+import sys
+
```
Comment (asridharan) [inline packages/dcos-net/extra/dcos-net-setup.py]: Can we add some comments here as to the purpose this file? Would be helpful for posterity.

File: packages/dcos-net/extra/dcos-net-observer.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+#!/opt/mesosphere/bin/python
+
+import http.client
+import sys
+import time
+
```
Comment (asridharan) [inline packages/dcos-net/extra/dcos-net-observer.py]: Can we add some comments here to explain the purpose of this file.

File: packages/adminrouter/extra/src/docs/api/nginx.master.yaml (hunk 2 above)
Comment (urbanserj) [reply packages/adminrouter/extra/src/docs/api/nginx.master.yaml]: Unfortunately, nginx doesn't support `-` symbol in literals :(

File: packages/dcos-net/extra/dcos-net-setup.py (hunk 5 above)
Comment (urbanserj) [reply packages/dcos-net/extra/dcos-net-setup.py]: https://github.com/mesosphere/dcos/blob/urbanserj/dcos-net/packages/dcos-net/extra/dcos-net-setup.py#L3-L10

File: packages/dcos-net/extra/dcos-net-observer.py (hunk 6 above)
Comment (urbanserj) [reply packages/dcos-net/extra/dcos-net-observer.py]: https://github.com/mesosphere/dcos/blob/urbanserj/dcos-net/packages/dcos-net/extra/dcos-net-observer.py#L3-L9

**Comment by asridharan** [comment]
@urbanserj do we need to re-base the enterprise build? We should try to get this committed at the earliest.

**Comment by urbanserj** [comment]
One test is failing [here](https://teamcity.mesosphere.io/viewLog.html?buildId=804168&buildTypeId=ClosedSource_MesosphereEnterpriseDcOs_IntegrationTests_DcOsCommonsSmokeTest&guest=1). I'm working on it.

**Comment by urbanserj** [comment]
@mesosphere-mergebot bump-ee

**Comment by urbanserj** [comment]
@mesosphere-mergebot bump-ee

**Comment by urbanserj** [comment]
It seems like disabled -> permissive upgrade test is broken: https://jira.mesosphere.com/browse/DCOS_OSS-1721.

**Comment by urbanserj** [comment]
@mesosphere-mergebot label [Ready For Review]

**Comment by asridharan** [comment]
@lingmann can you take another look at this. 
@jgehrcke @gpaul a look at the `bootstrap` changes would be helpful as well.

**Comment by asridharan** [comment]
@lingmann once we get the approvals on the enterprise PR, can we try putting this on its own train? Would like to see this move forward at the earliest.

**Comment by jgehrcke** [comment]
Ship-it was applied prematurely. See DCOS-19252.

**Comment by jgehrcke** [comment]
No bump-ee was performed after adding the last commits here. Doing that now. Also rebasing one more time.

**Comment by jgehrcke** [comment]
@mesosphere-mergebot bump-ee

**Comment by jgehrcke** [comment]
@gpaul can you please give this another sanity check?

File: gen/calc.py [hunk 7]
```diff
@@ -969,6 +970,7 @@ def calculate_fault_domain_detect_contents(fault_domain_detect_filename):
         'ui_banner_image_path': 'null',
         'ui_banner_dismissible': 'null',
         'dcos_net_rest_enable': "true",
+        'dcos_net_watchdog': "true",
```
[the remark this answers is not in the document]
Comment (gpaul) [reply gen/calc.py]: I notice that we've already lost consistency further down, please ignore.

File: packages/dcos-net/extra/dcos-net-watchdog.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        logging.erorr('Exception: {}'.sys.exc_info()[1])
+    return False
+
+
+def kill(name):
+    if os.getenv('DCOS_NET_WATCHDOG') == 'false':
+        logging.info('Killing is disabled')
+        return 0
+    logging.info('Killing %s', name)
+    r = subprocess.run(['/usr/bin/env', 'systemctl', 'kill', '-s', 9, name])
```
Comment (gpaul) [inline packages/dcos-net/extra/dcos-net-watchdog.py]: Perhaps we want `--kill-who=main` so that we don't kill a control process (ie., something running as `ExecStartPre=`) and incorrectly think we've killed the dcos_net service when really we've only killed bootstrap?

**Comment by gpaul** [comment]
Has anything been done regarding the observed issue where dcos-net-watchdog was killing bootstrap? Having the watchdog kill arbitrary ExecStartPre= commands which then roll forward to the next ExecStartPre= / ExecStart= clearly leads to undefined behaviour. 

If our recent experiments are correct, we will have a watchdog service running every 60s which will "kill -9 whatever ExecStartPre= is currently being executed by the dcos-net unit and proceed to the next ExecStartPre= / ExecStart=."

The current workaround of lazy-initializing the ZK client just hides the symptom. If dcos_net ever needs to do something with ZK (ie., the dcos_net bootstrapper is no longer "noop") the symptom will resurface and once more take several days to debug.

**Comment by gpaul** [comment]
Hence, my question regarding --kill-who=main

**Comment by urbanserj** [comment]
Hi @gpaul,

SIGTERM signal caused that issue. `systemctl kill` kills everything properly if it uses SIGKILL, so I added `-s 9` [here](https://github.com/mesosphere/dcos/blob/urbanserj/dcos-net/packages/dcos-net/extra/dcos-net-watchdog.py#L43) as a workaround. Thank you for mentioning `--kill-who=main` argument. I still need some time to check how it works. I'd like to be sure that it will kill main process and all child/forked processes in `dcos-net` cgroup as well.

File: packages/dcos-net/extra/dcos-net-watchdog.py (hunk 8 above)
Comment (urbanserj) [reply packages/dcos-net/extra/dcos-net-watchdog.py]: Yes! It would be great not to kill bootstrap process. I'll add that in the next dcos-net PR. Thank you!

TL;DR `--kill-who=main` doesn't work properly if `KillSignal=SIGKILL` is not specified.

I wrote a small daemon to check `systemctl kill` behavior.

```
import os, time, daemon
with daemon.DaemonContext():
    os.fork()
    time.sleep(100)
```

Here are some experiments (strace will trace forked process):

```
$ systemctl kill -s 9 --kill-who=main test
Failed to kill unit test.service: No main process to kill
```

```
$ systemctl kill -s 9 --kill-who=main test
$ strace -p xxx
rt_sigaction(SIGINT, …) = 0
rt_sigaction(SIGTERM, …) = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

```
$ systemctl kill -s 9 test
$ strace -p xxx
+++ killed by SIGKILL +++
```

```
$ systemctl cat test | grep KillSignal
KillSignal=SIGKILL
$ systemctl kill --kill-who=main test
  Process: yyy ExecStart=test.py (code=exited, status=1/FAILURE)
```

```
$ systemctl cat test | grep KillSignal
KillSignal=SIGKILL
$ systemctl kill -s 9 --kill-who=main test
  Process: yyy ExecStart=test.py (code=killed, signal=KILL)
$ strace -p xxx
+++ killed by SIGKILL +++
```

**Comment by urbanserj** [comment]
@gpaul Exactly. I tested `-s 9` without the lazy-loading patch. As it was mentioned [here](https://jira.mesosphere.com/browse/DCOS_OSS-1757?focusedCommentId=123490#comment-123490) I didn't revert my bootstrap patch just because it speeds up agent node initialization.

File: packages/dcos-net/extra/dcos-net-watchdog.py (hunk 8 above)
Comment (urbanserj) [reply packages/dcos-net/extra/dcos-net-watchdog.py]: https://github.com/mesosphere/dcos/blob/urbanserj/dcos-net-env/packages/dcos-net/extra/dcos-net.service#L7
https://github.com/mesosphere/dcos/blob/urbanserj/dcos-net-env/packages/dcos-net/extra/dcos-net-watchdog.py#L43-L47

**Comment by asridharan** [comment]
@lingmann can we build a separate train for this branch and move forward?

**Comment by orsenthil** [comment]
@lingmann  / @jgehrcke  - Just a note. This PR and the corresponding EE PR needs to land in a separate train. Please do not include this as part of another integration train. /cc: @mellenburg

**Comment by orsenthil** [comment]
Making a note here.  I triggered the train tests in this PR itself and I will work on taking this over the finish line.

**Comment by urbanserj** [comment]
@mesosphere-mergebot bump-ee

**Comment by urbanserj** [comment]
@mesosphere-mergebot bump-ee

**Comment by orsenthil** [comment]
@mesosphere-mergebot merge-it
